lm_eval/models/ggml.py

```diff
@@ -0,0 +1,96 @@
+import requests
+import logging
+from lm_eval.base import BaseLM
+from tqdm import tqdm
+from requests.exceptions import RequestException
+
+logger = logging.getLogger(__name__)
+
+def ggml_completion(base_url, **kwargs):
+    try:
+        response = requests.post(f"{base_url}/v1/completions", json=kwargs)
+        response.raise_for_status()
+        return response.json()
+    except RequestException as e:
+        print(f"RequestException: {e}")
+        return None
+
+class GGMLLM(BaseLM):
+    def __init__(self, base_url, truncate=False):
+        super().__init__()
+        self.base_url = base_url
+        self.truncate = truncate
+
+    def loglikelihood(self, requests):
+        res = []
+        for context, continuation in tqdm(requests):
+            response = ggml_completion(self.base_url, context=context, continuation=continuation)
+            if response and "choices" in response and response["choices"]:
+                choice = response["choices"][0]
+                logprobs = choice.get("logprobs")
+                logprob = logprobs["token_logprobs"][0] if logprobs and logprobs["token_logprobs"] else -1.2345
+                is_greedy = choice["finish_reason"] == "length"
+                res.append((logprob, is_greedy))
+            else:
+                logger.error(f"Invalid response for loglikelihood. Response: {response}")
+                assert False
+        return res
+
+    def greedy_until(self, requests):
+        if not requests:
+            return []
+
+        res = []
+        for request in tqdm(requests):
+            inp = request[0]
+            request_args = request[1]
+            until = request_args["until"]
+            response = ggml_completion(self.base_url, context=inp, stop=until)
+            if response and "text" in response:
+                generated_text = response["text"].strip()
+                res.append(generated_text)
+            else:
+                logger.error(f"Invalid response for greedy_until. Response: {response}")
```

Comment:
instead of just continuing, maybe either use a backoff timer and retry, or raise an exception when getting an invalid response?